sql/core/src/test/scala/org/apache/spark/sql/CollationExpressionWalkerSuite.scala

```diff
@@ -0,0 +1,652 @@
+
+    for (f <- funInfos.filter(f => !toSkip.contains(f.getName))) {
+      println("checking - " + f.getName)
+      val cl = Utils.classForName(f.getClassName)
+      val headConstructor = cl.getConstructors.head
+      val params = headConstructor.getParameters.map(p => p.getType)
+      val args = generateData(params.toSeq, Utf8Binary)
+      val expr = headConstructor.newInstance(args: _*).asInstanceOf[ExpectsInputTypes]
+      val inputTypes = expr.inputTypes
+
+      val inputDataUtf8Binary =
+        generateData(replaceExpressions(inputTypes, params.toSeq), Utf8Binary)
+      val instanceUtf8Binary =
+        headConstructor.newInstance(inputDataUtf8Binary: _*).asInstanceOf[Expression]
+      val inputDataLcase =
+        generateData(replaceExpressions(inputTypes, params.toSeq), Utf8BinaryLcase)
+      val instanceLcase = headConstructor.newInstance(inputDataLcase: _*).asInstanceOf[Expression]
+
+      val exceptionUtfBinary = {
+        try {
+          instanceUtf8Binary match {
+            case replaceable: RuntimeReplaceable =>
+              replaceable.replacement.eval(EmptyRow)
+            case _ =>
+              instanceUtf8Binary.eval(EmptyRow)
+          }
+          None
+        } catch {
+          case e: Throwable => Some(e)
+        }
+      }
+
+      val exceptionLcase = {
+        try {
+          instanceLcase match {
+            case replaceable: RuntimeReplaceable =>
+              replaceable.replacement.eval(EmptyRow)
+            case _ =>
+              instanceLcase.eval(EmptyRow)
+          }
+          None
+        } catch {
+          case e: Throwable =>
+            println(e.getMessage)
+            Some(e)
+        }
+      }
+
+      // Check that both cases either throw or pass
+      assert(exceptionUtfBinary.isDefined == exceptionLcase.isDefined)
+
+      if (exceptionUtfBinary.isEmpty) {
+        val resUtf8Binary = instanceUtf8Binary match {
+          case replaceable: RuntimeReplaceable =>
+            replaceable.replacement.eval(EmptyRow)
+          case _ =>
+            instanceUtf8Binary.eval(EmptyRow)
+        }
+        val resUtf8Lcase = instanceLcase match {
+          case replaceable: RuntimeReplaceable =>
+            replaceable.replacement.eval(EmptyRow)
+          case _ =>
+            instanceLcase.eval(EmptyRow)
+        }
+
+        val dt = instanceLcase.dataType
+
+        dt match {
+          case st if resUtf8Lcase != null && resUtf8Lcase != null && hasStringType(st) =>
+            // scalastyle:off caselocale
+            assert(resUtf8Binary.toString.toLowerCase === resUtf8Lcase.toString.toLowerCase)
+            // scalastyle:on caselocale
+          case _ =>
+            assert(resUtf8Lcase === resUtf8Binary)
+        }
+      }
+      else {
+        assert(exceptionUtfBinary.get.getClass == exceptionLcase.get.getClass)
+      }
+    }
+  }
+
+  test("SPARK-48280: Expression Walker for codeGen generation") {
```

Comment:
We usually do not test eval and codegen separately. See `ExpressionEvalHelper.checkEvaluation`

Reply:
This situation I would say is a bit different. We are not testing each expression with predeterminate solution. What we are doing is comparing UTF8_BINARY and UTF8_LCASE and trying to test whether they will return the same result. In this case it would be hard to use `ExpressionEvalHelper.checkEvaluation` in total. We could change the test to check whether both UTF8_BINARY and UTF8_LCASE are generating the same result in with codeGen and without, but I am not sure if there are other options than proposed PR for codeGen test